toolkit/scripts/sodiff/mariner-sodiff.sh

```diff
@@ -9,41 +9,93 @@ repo_file_path="$2"
 mariner_version="$3"
 sodiff_out_dir="$4"
 sodiff_log_file="${sodiff_out_dir}/sodiff.log"
+current_os=$(cat /etc/os-release | grep ^ID | cut -d'=' -f2)
+temporary_repo_file="file.repo"
 
 # Setup output dir
 mkdir -p "$sodiff_out_dir"
+# Empty the log file
+echo > "$sodiff_log_file"
 
-# Prepare mariner/ubuntu compatibility calls
+function makecache_with_common {
+    if [[ "$current_os" == mariner ]]; then
+        # Mariner uses DNF repoquery command
+        DNF_COMMAND=dnf
+        # Cache RPM metadata
+        >/dev/null dnf $common_options -y makecache
+    else
+        # Ubuntu uses repoquery command from yum-utils
+        DNF_COMMAND=
+        # Cache RPM metadata
+        # Ubuntu does not come with gpgcheck plugin for yum
+        >/dev/null yum $common_options -y --nogpgcheck makecache
+    fi
+}
+
+function log_to_file {
+    >>$sodiff_log_file echo -- $@
+}
+
+# Split the abrdiged repo file and process each repo separately
+# because if any of the repos is not accessible,
+# creating cache for the whole .repo file fails.
+# However, created cache combines , so creating multiple caches
+# Will not overwrite each other and any failures can be safely ignored
+
+# Flag to track running cache creation for the first match of repo header - skip since repo file not created yet
+firstmatch=1
+# Make common options point to the temporary repo file for now
+common_options="-c $temporary_repo_file --releasever $mariner_version"
+# IFS= prevents read from omitting leading/trailing whitespace
+# -r prevents read from interpreting backslash escapes
+while IFS= read -r line
+do
+    # If this is a start of a new repo definition
+    if grep -Eq "^\[[^]]+\]$" <<<"$line" ; then
+        # Do not run the operation on first match - repo file not filled yet
+        if [[ "$firstmatch" -eq 1 ]]; then
+            firstmatch=0
+        else
+            # Execute operation on the previous .repo file before proceeding with new one
+            makecache_with_common
+            echo ""
+        fi
+        log_to_file "Preparing cache for repo $line"
+        # Overwrite line to a repo file
+        printf '%s\n' "$line" > $temporary_repo_file
+    else
+        # Append line to a repo file
+        printf '%s\n' "$line" >> $temporary_repo_file
+    fi
+done < "$repo_file_path"
+# Run operation again for the final repo
+# In case repo file has only one repo definition, the command will
+# run twice on the same repo, but this is fine.
+makecache_with_common
```

Comment:
I think we can simplify a bit by splitting this into two step:
1. Separate the big repo file.
2. Generate cache.

```suggestion
# Splitting cumulative .repo file into single .repo files.
csplit -z -f 'repo' -b '_%d.repo' "$repo_file_path" '/^\[/' '{*}'

# Updating cache with each repo info separately.
for singe_repo_file in repo*.repo
do
    makecache_with_common "$single_repo_file"
done
```

With that change, you'd also need to update `makecache_with_common` to accept the path to the .repo file as an argument. I'd also suggest hard-coding the contents of `common_options` inside that function. Except for the .repo file path, of course.;)

Idea to use `csplit` taken from [here](https://unix.stackexchange.com/questions/263904/split-file-into-multiple-files-based-on-pattern).
Docs regarding the arguments [here](https://www.gnu.org/software/coreutils/manual/html_node/csplit-invocation.html).